src/libmongoc/src/mongoc/mongoc-collection.c

```diff
@@ -3503,3 +3508,35 @@ mongoc_collection_watch (const mongoc_collection_t *coll,
 {
    return _mongoc_change_stream_new_from_collection (coll, pipeline, opts);
 }
+
+bool
+mongoc_collection_set_timeout_ms (mongoc_collection_t *coll,
+                                  int64_t timeout_ms,
+                                  bson_error_t *error)
+{
+   BSON_ASSERT_PARAM (coll);
+
+   if (timeout_ms < -1) {
+      bson_set_error (error,
+                      MONGOC_ERROR_TIMEOUT,
+                      MONGOC_ERROR_TIMEOUT_INVALID,
+                      "timeoutMS must be a non-negative integer");
+      return false;
+   }
+
+   coll->timeout_ms = timeout_ms;
+
+   return true;
+}
+
+int64_t
+mongoc_collection_get_timeout_ms (const mongoc_collection_t *coll)
+{
+   BSON_ASSERT_PARAM (coll);
+
+   if (coll->timeout_ms == MONGOC_TIMEOUTMS_UNSET) {
+      return mongoc_client_get_timeout_ms (coll->client);
```

Comment:
I think this should return from the database.

Reply:
So, the collection doesn't keep track of its parent database.  If the parent db did have a timeoutMS set when it created this collection, it would have passed that value on to the collection, and coll->timeout_ms would not be UNSET.  Returning the client's timeout seemed like the best possible choice here, and still makes some attempt to follow the hierarchy.  

An earlier patch of this code review added in a parent db pointer onto the collection, so it could track changes in the parent db's timeout.  The feedback I got was that this wasn't ideal or necessary, since other things we inherit in a hierarchical fasion (like write_concern) function fine without it.

Reply:
Discussed offline. It seems preferable to be consistent with read concern / write concern / read preferences. Those are inherited when a collection struct is constructed, and will not change if the parent database / client's options change.

Reply:
Done.